### ISSUE TITLE
Closure for empty_data with empty string converted to null

### ISSUE THREAD

**Issue by Soullivaneuh** [symfony/symfony#25979]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | x.y.z

With this php sample:

```php
declare(strict_types=1);

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\Form\FormFactory;
use Symfony\Component\Form\FormInterface;
use Symfony\Component\Form\FormRegistry;
use Symfony\Component\Form\ResolvedFormTypeFactory;

require_once __DIR__.'/vendor/autoload.php';

class ThingType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('name', null, [
                'empty_data' => '',
            ])
            ->add('closure', null, [
                'empty_data' => function (FormInterface $form) {
                    return '';
                },
            ])
        ;
    }
}

$formFactory = new FormFactory(new FormRegistry([], new ResolvedFormTypeFactory()));
$form = $formFactory->create(ThingType::class);
$form->submit([]);
dump($form->getData());
```

You have:

```
sullivan@c8dc0d9ca068:/code$ php test.php 
array:2 [
  "name" => ""
  "closure" => null
]
```

Both should be empty strings.

**Comment by jvasseur** [comment]
That's normal. The `empty_data` options is supposed to contains data in the "view" form, this means it gets transformed back to the "model" form afterward. That's why you get `null` when the empty data is `''`.

**Comment by yceruto** [comment]
> But `'empty_data' => ''` giving `''` and not `null` is strange and looks like a bug.

That's understood for text fields, see [Issue](https://github.com/symfony/symfony/issues/5906) and [PR](https://github.com/symfony/symfony/pull/18357). Btw, it isn't documented in  http://symfony.com/doc/current/reference/forms/types/text.html#empty-data.

> "closure" => null

The inconsistency comes from here, where the view transformer is not added:
https://github.com/symfony/symfony/blob/b66afe62e86dc2d50092489ea557daee1e6d723f/src/Symfony/Component/Form/Extension/Core/Type/TextType.php#L28-L30

**Comment by Soullivaneuh** [comment]
> But `'empty_data' => ''` giving `''` and not `null` is strange and looks like a bug.

@jvasseur Here I don't understand. Why `empty_data` should return `null` if I tell it to return `''`?

@yceruto I think you indeed found the issue location. But why not simply add the view transformer even if the `empty_data` is a closure? :thinking:

**Comment by Soullivaneuh** [comment]
> So, maybe a `SUBMIT` listener and do the transformation there?

@yceruto I can't bind a listener on the `SUBMIT` event because it crash before triggering on this event.

I tried on `PRE_SUBMIT`, with a simple `dump`:

```php
use Symfony\Component\Form\AbstractTypeExtension;
use Symfony\Component\Form\Extension\Core\Type\TextType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\Form\FormEvent;
use Symfony\Component\Form\FormEvents;

final class TextTypeExtension extends AbstractTypeExtension
{
    public function getExtendedType(): string
    {
        return TextType::class;
    }

    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder->addEventListener(FormEvents::PRE_SUBMIT, [$this, 'onSubmit']);
    }

    public function onSubmit(FormEvent $event)
    {
        dump($event);
    }
}
```

But here the data is correct:

```
FormEvent {#3266 ▼
  -form: Form {#3281 ▼
    -config: FormBuilder {#3279 ▶}
    -parent: Form {#3290 ▶}
    -children: OrderedHashMap {#3274 ▶}
    -errors: []
    -submitted: false
    -clickedButton: null
    -modelData: ""
    -normData: ""
    -viewData: ""
    -extraData: []
    -transformationFailure: null
    -defaultDataSet: true
    -lockSetData: false
  }
  #data: ""
  -propagationStopped: false
}
```

So I don't really see what I can do in this case.

**Comment by jvasseur** [comment]
@Soullivaneuh because of this code : https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Form.php#L1128-L1130 (if you don't have any view transformer you get a "default" one that converts `''` to `null`).

I don't think it's a good thing but it's would be hard to change without BC break. But you can easily change that by adding an "identity" view transformer.

@yceruto I don't understand the logic behind #18357 it's a documented feature that `empty_data` is in view form and that the view to norm transform `''` to `null` and this PR breaks this assumption (and only in one specific case which is broken as seen here).

**Comment by Soullivaneuh** [comment]
> But you can easily change that by adding an "identity" view transformer.

Maybe could you add a sample? I would like to check the data_class and the properties like I did in https://github.com/symfony/symfony/issues/25977#issuecomment-361758227 but I can't access if from the event listener.

> I don't understand the logic behind #18357

Well, bug or not, it should be clarified in this case.

**Comment by jvasseur** [comment]
> Maybe could you add a sample? I would like to check the data_class and the properties like I did in https://github.com/symfony/symfony/issues/25977#issuecomment-361758227 but I can't access if from the event listener.


You just need to add
```php
$builder->get('name')->addViewTransformer(new class() implements DataTransformerInterface {
    public function transform($value)
    {
        return $value;
    }

    public function reverseTransform($value)
    {
        return $value;
    }
});
```
to your code to disable the default empty string to null transformation.

You can't do it inside an event listener but you could easily do it in the buildForm method.

**Comment by Soullivaneuh** [comment]
@jvasseur Yes I can indeed do it for this very special case. But I want to make it generic with an extension as done in https://github.com/symfony/symfony/issues/25977#issuecomment-361758227.

Plus, if the property is nullable, I would like to keep the `null` transformation.

**Comment by Soullivaneuh** [comment]
Well I spoke to fast. This workaround as an extension work. See my update: https://github.com/symfony/symfony/issues/25977#issuecomment-361758227

Still, is it a bug for me. If `'empty_data' => ''` returns `''`, I don't see why a closure should not.

**Comment by Soullivaneuh** [comment]
@jvasseur Your trick does not work for repeated types:

```php
class DomainFtpAccountPasswordType extends AbstractType
{
    /**
     * {@inheritdoc}
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder->add('password', RepeatedType::class, [
            'translation_domain' => 'global',
            'type' => PasswordType::class,
            'invalid_message' => 'global.password_mismatch',
            'required' => true,
            'first_options' => ['label' => 'form.new_password', 'attr' => ['class' => 'strength']],
            'second_options' => ['label' => 'form.confirmation'],
        ]);
    }
}
```

Will submitted empty data, the `FtpAccount::setPassword` is call with a `null` value. Any idea for a workaround?

**Comment by jvasseur** [comment]
@Soullivaneuh you have to set the DataTransformer on the childs of the RepeatedType, something like this should work:
```php
$builder->get('password')->get('first')->addViewTransformer();
$builder->get('password')->get('second')->addViewTransformer();
```

**Comment by jvasseur** [comment]
@Soullivaneuh you can add a second form extension that extends the repeated type to add the transformer to both the `first` and `second` field.

**Comment by Soullivaneuh** [comment]
@jvasseur Not needed, I found a way to handle all of them. Please see: https://github.com/symfony/symfony/issues/25977#issuecomment-389901913

**Comment by soullivaneuh** [comment]
@xabbuh The solution proposed by @jvasseur is not working because the `ReapetedType` has a dedicated view transformer that also turn empty arrays to null, even for array with empty strings.

I found a workaround posted https://github.com/symfony/symfony/issues/25977#issuecomment-999739106, but I am quite sure some vendor simplification would be welcomed here.
